### PR TITLE
feat: Add translation key prefixes for upload and download

### PR DESCRIPTION
**Pull request by malvarezphrase**
Purpose: Provide documentation for the new params on import and export.

**Pushing**:
- provide a prefix with `translation_key_prefix` which will be prepended to the name of the imported translation keys
- provide `<file_name>` as the `translation_key_prefix` to use the file path of the pushed file as a prefix

**Pulling**:
- provide a prefix with `translation_key_prefix` which will be subtracted from the name of the exported translation keys. Warning: this may create duplicate key names if other keys share the same name after the prefix is removed
- provide `<file_name>` as the `translation_key_prefix` to use the file path of the pulled file as a prefix
- set the option `use_translation_key_prefix_as_filter` to `true` for the prefix to work as a filter. In this case only translation keys including the provided prefix in their name will be exported

Related tickets:
https://phrase.atlassian.net/browse/STRINGS-374
https://phrase.atlassian.net/browse/STRINGS-439

### CODE REVIEW

**Comment by jablan** [comment]
hey @malvarezphrase can you please explain the logic with the truncating file name? I don't get it.

**Comment by malvarezphrase** [comment]
> hey @malvarezphrase can you please explain the logic with the truncating file name? I don't get it.

@jablan The reason for the truncate of the file name is that we want to use the `idx_translation_keys_on_name_project_deleted` index on translation keys when the prefix is used as a filter. Therefore we only allow it to be maximum 255 characters long.

Now looking at the other usages of the prefix, is it also possible to add a limit on `create.yaml` and `download.yaml`? Or will phrase respond with a validation error?

**Comment by jablan** [comment]
@malvarezphrase sorry, I need a bit more context. where does `<file_path>` come from? is that a "magic" prefix working similar to `<locale_code>` and such? is it explained somewhere?

**Comment by jablan** [comment]
> Where would I put it?

@malvarezphrase I would start with the ticket description :smile: 

if it can be specified only through the `phrase.yml`, probably we should document it in https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/phrase.json
but I don't see other magic placeholders either, so perhaps it's a good opportunity to add them all?

regarding the truncating, if I understand you correctly, we don't support prefixes longer than 255 chars? in that case we should probably error out rather then truncating them?

**Comment by malvarezphrase** [comment]
@jablan Thanks a lot for the feedback! I already updated the ticket descriptions and also the description of this PR.

Are the changes to https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/phrase.json being done manually? If so, I will prepare a PR for it.

> in that case we should probably error out rather than truncating them?
How do errors on the CLI usually work? The validation for the length is already in place inside phrase.

**Comment by jablan** [comment]
perhaps then just passing whatever prefix is provided (either explicitly by the user or by using the filename) will be enough. is the error clear to the user?

as for json schema, perhaps you could open a small ticket, I see couple other things that could be fixed there, so we will do it all together.

just remember, we also have https://support.phrase.com/hc/en-us/articles/5808300599068-Using-the-CLI-Strings where we might want to describe this feature?

**Comment by malvarezphrase** [comment]
@jablan & @theSoenke after changing to not truncate the length of the prefix and updating the description [here](https://github.com/phrase/schemastore/pull/1), is there anything else missing?

File: doc/compiled.json [hunk 1]
```diff
@@ -8775,6 +8775,24 @@
...
+            "name": "translation_key_prefix",
+            "in": "query",
+            "schema": {
+              "type": "string"
+            }
+          },
+          {
+            "description": "Only download translation keys containing the specified prefix, and remove the prefix from the generated file.",
+            "example": null,
+            "name": "use_translation_key_prefix_as_filter",
```
Comment (theSoenke) [inline doc/compiled.json:8790]: This still seems like a really long name to me that's hard to remember. Maybe `filter_by_prefix` instead?

Comment (theSoenke) [reply doc/compiled.json:8790]: I'm also still not sure it's good to have this option to disable it. It also violates our principle that imported files can be uploaded as is again, when there's a conflict. And also there might be formats where this is causing issues during rendering

Comment (malvarezphrase) [reply doc/compiled.json:8790]: I can certainly change the name of the parameter ^^

> I'm also still not sure it's good to have this option to disable it. It also violates our principle that imported files can be uploaded as is again, when there's a conflict. And also there might be formats where this is causing issues during rendering
Do you suggest to by default use the prefix as a filter? In that case, how would you name a second parameter if the filtering is not wanted?
Regarding the issues: Maybe it's enough to warn the users. Let me ask our PM.

Comment (malvarezphrase) [reply doc/compiled.json]: `use_translation_key_prefix_as_filter` was renamed to `filter_by_prefix` 👍 
See also here: https://github.com/phrase/phrase/pull/13996

File: clients/cli/cmd/internal/pull_target.go [hunk 2]
```diff
@@ -190,6 +191,9 @@ func TargetsFromConfig(config phrase.Config) (Targets, error) {
 		if target.FileFormat == "" {
 			target.FileFormat = fileFormat
 		}
+		if target.Params.TranslationKeyPrefix.Value() == "<file_path>" {
+			target.Params.TranslationKeyPrefix = optional.NewString(target.File)
```
[the remark this answers is not in the document]
Comment (theSoenke) [reply clients/cli/cmd/internal/pull_target.go:195]: That's important! Really good catch and it does still contain placeholders at this point

**Comment by theSoenke** [comment]
@jablan @malvarezphrase I've changed the `file_path` placeholder to make it work with other placeholders. It's also now possible to use the placeholder as part of the prefix like `<file_path>-foo`. Having the ability to separate the path from the key name by characters seems useful when I was testing it

**Review by jablan — APPROVED** [review]
:ok_hand:

**Pull request merged**